internal/commands/root/middleware.go

```diff
@@ -0,0 +1,67 @@
+package root
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/virtual-kubelet/virtual-kubelet/log"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	kubeletserver "k8s.io/kubernetes/pkg/kubelet/server"
+)
+
+// AuthMiddleware contains all methods required by the auth filters
+type AuthMiddleware interface {
+	AuthFilter(h http.HandlerFunc) http.HandlerFunc
+}
+
+// KubeletAuthMiddleware is the struct to implement middleware
+type KubeletAuthMiddleware struct {
+	auth kubeletserver.AuthInterface
+	ctx  context.Context
+}
+
+// NewKubeletKubeletAuthMiddleware initiate an instance for AuthMiddleware
+func NewKubeletKubeletAuthMiddleware(ctx context.Context, auth kubeletserver.AuthInterface) AuthMiddleware {
+	return KubeletAuthMiddleware{auth: auth, ctx: ctx}
+}
+
+// AuthFilter is the middleware to authenticate & authorize the request
+func (m KubeletAuthMiddleware) AuthFilter(h http.HandlerFunc) http.HandlerFunc {
+	return http.HandlerFunc(func(resp http.ResponseWriter, req *http.Request) {
+		info, ok, err := m.auth.AuthenticateRequest(req)
+		if err != nil {
+			log.G(m.ctx).Infof("Unauthorized, err: %s", err)
+			resp.WriteHeader(http.StatusUnauthorized)
+			resp.Write([]byte("Unauthorized"))
+
+			return
+		}
+		if !ok {
+			log.G(m.ctx).Infof("Unauthorized, ok: %t", ok)
```

Comment:
`kubectl exec` got Unauthorized here because ok is false. Seems token is not used when apiserver talks to virutal kubelete http server.

Reply:
we should enable tls to solve this issue.